base/src/main/java/com/mindera/skeletoid/widgets/tooltip/PopupTooltip.kt

```diff
@@ -0,0 +1,466 @@
+        private var showDuration: Long = 3500L,
+        private var verticalMargin: Int = 0,
+        private var horizontalMargin: Int = 0,
+        private var horizontalMarginClipped: Int = 0,
+        private var verticalMarginClipped: Int = 0,
+        private var insideTouchDismissible: Boolean = true,
+        private var outsideTouchDismissible: Boolean = true,
+        private var respectScreenBounds: Boolean = false,
+        private var modal: Boolean = true,
+        @StyleRes private var animation: Int? = null,
+        //Tooltip arrow
+        private var arrowWidth: Int = ARROW_WIDTH,
+        private var arrowHeight: Int = ARROW_HEIGHT,
+        private var arrow: Boolean = false,
+        @DrawableRes private var arrowDrawable: Int? = null,
+        @ColorRes private var arrowStrokeColor: Int = android.R.color.white,
+        @IdRes private var arrowBackgroundId: Int? = null,
+        @IdRes private var arrowStrokeId: Int? = null
+) : PopupWindow() {
+
+    companion object {
+
+        /**
+         * The LOGGING TAG. Epic.
+         */
+        const val TAG = "PopupTooltip"
+
+        /**
+         * Saves a reference to our currently displayed tooltip. This is here so we can control when
+         * we dismiss/show it.
+         */
+        private var tooltip: PopupWindow? = null
+
+        /**
+         * The estimated height of our popup in dips (density-independent pixels).
+         * This assumes the popup will have two lines of text with the text sizes
+         * specified in [layout].
+         */
+        const val ESTIMATED_HEIGHT_DIPS = 46
+
+        /**
+         * The default width value of the tooltip if none is specified, in dips (density-independent pixels).
+         */
+        const val TOOLTIP_WIDTH = 250
+
+        /**
+         * The default width value of the tooltip arrow if none is specified, in dips (density-independent pixels).
+         */
+        const val ARROW_WIDTH = 10
+
+        /**
+         * The default height value of the tooltip arrow if none is specified, in dips (density-independent pixels).
+         */
+        const val ARROW_HEIGHT = 10
+
+        /**
+         * The default elevation of the tooltip, in dips (density-independent pixels).
+         */
+        const val DEFAULT_ELEVATION = 5
+
+        /**
+         * At the moment this is hardcoded here. In the future the idea is for this
+         * to be changeable.
+         */
+        const val GRAVITY = Gravity.NO_GRAVITY
+
+        /**
+         * Sets the tooltip as null after showDuration amount of time. This is only used
+         * when we set the tooltip as definite.
+         */
+        private val handler = Handler()
+
+        /**
+         * Clears the current reference to the tooltip and dismisses if it is present.
+         */
+        fun clearTooltipIfNotNull() {
+            tooltip?.apply {
+                if (isShowing) dismiss()
+                tooltip = null
+            }
```

Comment:
I find this hard to read. can we find a better idiomatic way to accomplish that

maybe
```
private fun PopupWindow.dismissIfShowing() {
    if (isShowing) dismiss()
}

fun clearTooltip() {
    tooltip?.dismissIfShowing()
    tooltip = null
}
```

Reply:
Sure, changed 👍

Reply:
Sure 👍

Reply:
@ruialmeida51 Please "resolve conversation" on all comments fixed ;-)